engine.py

```diff
@@ -112,42 +113,52 @@ def execute_command(self, cmd_key, args):
         if cb_var_args:
             have_expected_args = (len(args) >= len(cb_arg_list))
         else:
-            have_expected_args = (len(args) == len(cb_arg_list)) 
-        
+            have_expected_args = (len(args) == len(cb_arg_list))
+
         if not have_expected_args:
             expected_args = list(cb_arg_list)
             if cb_var_args:
                 expected_args.append("*%s" % cb_var_args)
             raise TankError("Cannot run command! Expected command arguments (%s)" % ", ".join(expected_args))
-        
+
         if not self._has_qt:
             # QT not available - just run the command straight
             return cb(*args)
         else:
             from tank.platform.qt import QtCore, QtGui
-            
-            # we got QT capabilities. Start a QT app and fire the command into the app
-            tk_shell = self.import_module("tk_shell")
-            t = tk_shell.Task(self, cb, args)
-            
-            # start up our QApp now, if none is already running
-            qt_application = None
+
+            # start up our QApp now
+            qt_application = QtGui.QApplication([])
+            qt_application.setWindowIcon(QtGui.QIcon(self.icon_256))
+
             if not QtGui.qApp:
                 # We need to clear Qt library paths on Linux if KDE is the active environment.
                 # This resolves issues with mismatched Qt libraries between the OS and the
                 # application being launched if it is a DCC that comes with a bundled Qt.
                 if sys.platform == "linux2" and os.environ.get("KDE_FULL_SESSION") is not None:
                     QtGui.QApplication.setLibraryPaths([])
 
-                qt_application = QtGui.QApplication([])
-                qt_application.setWindowIcon(QtGui.QIcon(self.icon_256))
-                self._initialize_dark_look_and_feel()
+            # we got QT capabilities. Start a QT app and fire the command into the app
+            tk_shell = self.import_module("tk_shell")
+            t = tk_shell.Task(self, cb, args)
+            self._initialize_dark_look_and_feel()
+
+            # when the QApp starts, initialize our task code
+            QtCore.QTimer.singleShot(0, t.run_command)
+
+            # and ask the main app to exit when the task emits its finished signal
+            t.finished.connect(qt_application.quit)
+
+            # start the application loop. This will block the process until the task
```

Comment:
There is a lot of code change here and it's not clear to me why we're doing it. Could you explain?